Cannot built treelite4j.dll on Windows

I was trying to install treelite4j, when implementing  `cmake --build . --config Release` I got :

The complete information is as follows:
```D:\treelite\build (mainline -> origin)  cmake --build . --config Release
Consolidate compiler generated dependencies of target objtreelite_common
[ 13%] Built target objtreelite_common
Consolidate compiler generated dependencies of target objtreelite
[ 83%] Built target objtreelite
[ 86%] Built target treelite
Consolidate compiler generated dependencies of target objtreelite_runtime
[ 89%] Building CXX object src/CMakeFiles/objtreelite_runtime.dir/predictor/predictor.cc.obj
D:\treelite\src\predictor\predictor.cc: In member function 'void* treelite::predictor::SharedLibrary::LoadFunction(const char*) const':
D:\treelite\src\predictor\predictor.cc:184:54: error: invalid static_cast from type 'FARPROC' {aka 'long long int (*)()'} to type 'void*'
   void* func_handle = static_cast<void*>(_func_handle)
                                                      ^
In file included from D:\treelite\src\predictor\predictor.cc:20:
D:\treelite\src\predictor\thread_pool/thread_pool.h: In instantiation of 'void treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::SetAffinity() [with InputToken = {anonymous}::InputToken; OutputToken = {anonymous}::OutputToken; TaskContext = treelite::predictor::Predictor]':
D:\treelite\src\predictor\thread_pool/thread_pool.h:55:7:   required from 'treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::ThreadPool(int, const TaskContext*, treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::TaskFunc) [with InputToken = {anonymous}::InputToken; OutputToken = {anonymous}::OutputToken; TaskContext = treelite::predictor::Predictor; treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::TaskFunc = void (*)(treelite::predictor::SpscQueue<{anonymous}::InputToken>*, treelite::predictor::SpscQueue<{anonymous}::OutputToken>*, const treelite::predictor::Predictor*)]'
D:\treelite\src\predictor\predictor.cc:375:6:   required from here
D:\treelite\src\predictor\thread_pool/thread_pool.h:88:53: warning: invalid conversion from 'std::thread::native_handle_type' {aka 'long long unsigned int'} to 'HANDLE' {aka 'void*'} [-fpermissive]
       SetThreadAffinityMask(thread_[i].native_handle(), (1ULL << core_id));
                             ~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from C:/PROGRA~2/mingw64/x86_64-w64-mingw32/include/windows.h:70,
                 from D:\treelite\src\predictor\thread_pool/thread_pool.h:15,
                 from D:\treelite\src\predictor\predictor.cc:20:
C:/PROGRA~2/mingw64/x86_64-w64-mingw32/include/winbase.h:1162:61: note:   initializing argument 1 of 'DWORD_PTR SetThreadAffinityMask(HANDLE, DWORD_PTR)'
   WINBASEAPI DWORD_PTR WINAPI SetThreadAffinityMask (HANDLE hThread, DWORD_PTR dwThreadAffinityMask);
                                                      ~~~~~~~^~~~~~~
mingw32-make.exe[2]: *** [src\CMakeFiles\objtreelite_runtime.dir\build.make:91: src/CMakeFiles/objtreelite_runtime.dir/predictor/predictor.cc.obj] Error 1
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:247: src/CMakeFiles/objtreelite_runtime.dir/all] Error 2
mingw32-make.exe: *** [Makefile:135: all] Error 2
```
Both -G"Visual Studio 16 2019" and -G"MinGW Makefiles" had been used as `maybe_generator` and both failed.

I've updated the code and now trying to build treelite4j 2.2.2. However,  I still failed when implementing `cmake --build . --config Release`. I got: `subprocess.CalledProcessError: Command 'cmake --build . --config Release' returned non-zero exit status 2.`

The complete information is as follows:
```mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
[  2%] Building CXX object src/CMakeFiles/objtreelite.dir/c_api/c_api.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/c_api/c_api.cc.obj -MF CMakeFiles\objtreelite.dir\c_api\c_api.cc.obj.d -o CMakeFiles\objtreelite.dir\c_api\c_api.cc.obj -c D:\tre
elite\src\c_api\c_api.cc
[  5%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/build.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/build.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\build.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\ast\b
uild.cc.obj -c D:\treelite\src\compiler\ast\build.cc
[  8%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/dump.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/dump.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\dump.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\ast\dum
p.cc.obj -c D:\treelite\src\compiler\ast\dump.cc
[ 10%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/fold_code.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/fold_code.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\fold_code.cc.obj.d -o CMakeFiles\objtreelite.dir\compil
er\ast\fold_code.cc.obj -c D:\treelite\src\compiler\ast\fold_code.cc
[ 13%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/is_categorical_array.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/is_categorical_array.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\is_categorical_array.cc.obj.d -o CMakeFiles\
objtreelite.dir\compiler\ast\is_categorical_array.cc.obj -c D:\treelite\src\compiler\ast\is_categorical_array.cc
[ 16%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/load_data_counts.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/load_data_counts.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\load_data_counts.cc.obj.d -o CMakeFiles\objtreel
ite.dir\compiler\ast\load_data_counts.cc.obj -c D:\treelite\src\compiler\ast\load_data_counts.cc
[ 18%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/quantize.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/quantize.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\quantize.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler
\ast\quantize.cc.obj -c D:\treelite\src\compiler\ast\quantize.cc
[ 21%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast/split.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast/split.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast\split.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\ast\s
plit.cc.obj -c D:\treelite\src\compiler\ast\split.cc
[ 24%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/elf/elf_formatter.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/elf/elf_formatter.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\elf\elf_formatter.cc.obj.d -o CMakeFiles\objtreelite.di
r\compiler\elf\elf_formatter.cc.obj -c D:\treelite\src\compiler\elf\elf_formatter.cc
[ 27%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/ast_native.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/ast_native.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\ast_native.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\ast
_native.cc.obj -c D:\treelite\src\compiler\ast_native.cc
[ 29%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/compiler.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/compiler.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\compiler.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\compile
r.cc.obj -c D:\treelite\src\compiler\compiler.cc
[ 32%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/failsafe.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/failsafe.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\failsafe.cc.obj.d -o CMakeFiles\objtreelite.dir\compiler\failsaf
e.cc.obj -c D:\treelite\src\compiler\failsafe.cc
[ 35%] Building CXX object src/CMakeFiles/objtreelite.dir/compiler/pred_transform.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/compiler/pred_transform.cc.obj -MF CMakeFiles\objtreelite.dir\compiler\pred_transform.cc.obj.d -o CMakeFiles\objtreelite.dir\comp
iler\pred_transform.cc.obj -c D:\treelite\src\compiler\pred_transform.cc
[ 37%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/builder.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/builder.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\builder.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend\builder.c
c.obj -c D:\treelite\src\frontend\builder.cc
[ 40%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/lightgbm.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/lightgbm.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\lightgbm.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend\lightgb
m.cc.obj -c D:\treelite\src\frontend\lightgbm.cc
[ 43%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/sklearn.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/sklearn.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\sklearn.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend\sklearn.c
c.obj -c D:\treelite\src\frontend\sklearn.cc
[ 45%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/xgboost.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/xgboost.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\xgboost.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend\xgboost.c
c.obj -c D:\treelite\src\frontend\xgboost.cc
[ 48%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/xgboost_json.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/xgboost_json.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\xgboost_json.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend
\xgboost_json.cc.obj -c D:\treelite\src\frontend\xgboost_json.cc
[ 51%] Building CXX object src/CMakeFiles/objtreelite.dir/frontend/xgboost_util.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/frontend/xgboost_util.cc.obj -MF CMakeFiles\objtreelite.dir\frontend\xgboost_util.cc.obj.d -o CMakeFiles\objtreelite.dir\frontend
\xgboost_util.cc.obj -c D:\treelite\src\frontend\xgboost_util.cc
[ 54%] Building CXX object src/CMakeFiles/objtreelite.dir/gtil/predict.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/gtil/predict.cc.obj -MF CMakeFiles\objtreelite.dir\gtil\predict.cc.obj.d -o CMakeFiles\objtreelite.dir\gtil\predict.cc.obj -c D:\
treelite\src\gtil\predict.cc
[ 56%] Building CXX object src/CMakeFiles/objtreelite.dir/gtil/pred_transform.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/gtil/pred_transform.cc.obj -MF CMakeFiles\objtreelite.dir\gtil\pred_transform.cc.obj.d -o CMakeFiles\objtreelite.dir\gtil\pred_tr
ansform.cc.obj -c D:\treelite\src\gtil\pred_transform.cc
[ 59%] Building CXX object src/CMakeFiles/objtreelite.dir/annotator.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/annotator.cc.obj -MF CMakeFiles\objtreelite.dir\annotator.cc.obj.d -o CMakeFiles\objtreelite.dir\annotator.cc.obj -c D:\treelite\
src\annotator.cc
[ 62%] Building CXX object src/CMakeFiles/objtreelite.dir/filesystem.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/filesystem.cc.obj -MF CMakeFiles\objtreelite.dir\filesystem.cc.obj.d -o CMakeFiles\objtreelite.dir\filesystem.cc.obj -c D:\treeli
te\src\filesystem.cc
[ 64%] Building CXX object src/CMakeFiles/objtreelite.dir/optable.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/optable.cc.obj -MF CMakeFiles\objtreelite.dir\optable.cc.obj.d -o CMakeFiles\objtreelite.dir\optable.cc.obj -c D:\treelite\src\op
table.cc
[ 67%] Building CXX object src/CMakeFiles/objtreelite.dir/serializer.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/serializer.cc.obj -MF CMakeFiles\objtreelite.dir\serializer.cc.obj.d -o CMakeFiles\objtreelite.dir\serializer.cc.obj -c D:\treeli
te\src\serializer.cc
[ 70%] Building CXX object src/CMakeFiles/objtreelite.dir/json_serializer.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DFMT_HEADER_ONLY=1 -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fo
penmp -std=gnu++14 -MD -MT src/CMakeFiles/objtreelite.dir/json_serializer.cc.obj -MF CMakeFiles\objtreelite.dir\json_serializer.cc.obj.d -o CMakeFiles\objtreelite.dir\json_serializer.cc.o
bj -c D:\treelite\src\json_serializer.cc
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
[ 70%] Built target objtreelite
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f src\CMakeFiles\objtreelite_common.dir\build.make src/CMakeFiles/objtreelite_common.dir/depend
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
"C:\Program Files\CMake\bin\cmake.exe" -E cmake_depends "MinGW Makefiles" D:\treelite D:\treelite\src D:\treelite\build D:\treelite\build\src D:\treelite\build\src\CMakeFiles\objtreelite_
common.dir\DependInfo.cmake --color=
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f src\CMakeFiles\objtreelite_common.dir\build.make src/CMakeFiles/objtreelite_common.dir/build
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
[ 72%] Building CXX object src/CMakeFiles/objtreelite_common.dir/c_api/c_api_common.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_common.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=gn
u++14 -MD -MT src/CMakeFiles/objtreelite_common.dir/c_api/c_api_common.cc.obj -MF CMakeFiles\objtreelite_common.dir\c_api\c_api_common.cc.obj.d -o CMakeFiles\objtreelite_common.dir\c_api\
c_api_common.cc.obj -c D:\treelite\src\c_api\c_api_common.cc
[ 75%] Building CXX object src/CMakeFiles/objtreelite_common.dir/c_api/c_api_error.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_common.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=gn
u++14 -MD -MT src/CMakeFiles/objtreelite_common.dir/c_api/c_api_error.cc.obj -MF CMakeFiles\objtreelite_common.dir\c_api\c_api_error.cc.obj.d -o CMakeFiles\objtreelite_common.dir\c_api\c_
api_error.cc.obj -c D:\treelite\src\c_api\c_api_error.cc
[ 78%] Building CXX object src/CMakeFiles/objtreelite_common.dir/data.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_common.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=gn
u++14 -MD -MT src/CMakeFiles/objtreelite_common.dir/data.cc.obj -MF CMakeFiles\objtreelite_common.dir\data.cc.obj.d -o CMakeFiles\objtreelite_common.dir\data.cc.obj -c D:\treelite\src\dat
a.cc
[ 81%] Building CXX object src/CMakeFiles/objtreelite_common.dir/logging.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_common.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=gn
u++14 -MD -MT src/CMakeFiles/objtreelite_common.dir/logging.cc.obj -MF CMakeFiles\objtreelite_common.dir\logging.cc.obj.d -o CMakeFiles\objtreelite_common.dir\logging.cc.obj -c D:\treelit
e\src\logging.cc
[ 83%] Building CXX object src/CMakeFiles/objtreelite_common.dir/typeinfo.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_common.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=gn
u++14 -MD -MT src/CMakeFiles/objtreelite_common.dir/typeinfo.cc.obj -MF CMakeFiles\objtreelite_common.dir\typeinfo.cc.obj.d -o CMakeFiles\objtreelite_common.dir\typeinfo.cc.obj -c D:\tree
lite\src\typeinfo.cc
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
[ 83%] Built target objtreelite_common
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f CMakeFiles\treelite.dir\build.make CMakeFiles/treelite.dir/depend
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
"C:\Program Files\CMake\bin\cmake.exe" -E cmake_depends "MinGW Makefiles" D:\treelite D:\treelite D:\treelite\build D:\treelite\build D:\treelite\build\CMakeFiles\treelite.dir\DependInfo.
cmake --color=
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f CMakeFiles\treelite.dir\build.make CMakeFiles/treelite.dir/build
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
[ 86%] Linking CXX shared library libtreelite.dll
"C:\Program Files\CMake\bin\cmake.exe" -E cmake_link_script CMakeFiles\treelite.dir\link.txt --verbose=1
"C:\Program Files\CMake\bin\cmake.exe" -E rm -f CMakeFiles\treelite.dir/objects.a
C:\PROGRA~2\mingw64\bin\ar.exe qc CMakeFiles\treelite.dir/objects.a @CMakeFiles\treelite.dir\objects1.rsp
C:\PROGRA~2\mingw64\bin\G__~1.EXE -O3 -DNDEBUG -shared -o libtreelite.dll -Wl,--out-implib,libtreelite.dll.a -Wl,--major-image-version,0,--minor-image-version,0 -Wl,--whole-archive CMakeF
iles\treelite.dir/objects.a -Wl,--no-whole-archive @CMakeFiles\treelite.dir\linklibs.rsp
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
[ 86%] Built target treelite
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f src\CMakeFiles\objtreelite_runtime.dir\build.make src/CMakeFiles/objtreelite_runtime.dir/depend
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
"C:\Program Files\CMake\bin\cmake.exe" -E cmake_depends "MinGW Makefiles" D:\treelite D:\treelite\src D:\treelite\build D:\treelite\build\src D:\treelite\build\src\CMakeFiles\objtreelite_
runtime.dir\DependInfo.cmake --color=
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
C:/PROGRA~2/mingw64/bin/mingw32-make.exe  -f src\CMakeFiles\objtreelite_runtime.dir\build.make src/CMakeFiles/objtreelite_runtime.dir/build
mingw32-make.exe[2]: Entering directory 'D:/treelite/build'
[ 89%] Building CXX object src/CMakeFiles/objtreelite_runtime.dir/c_api/c_api_runtime.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_runtime.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=g
nu++14 -MD -MT src/CMakeFiles/objtreelite_runtime.dir/c_api/c_api_runtime.cc.obj -MF CMakeFiles\objtreelite_runtime.dir\c_api\c_api_runtime.cc.obj.d -o CMakeFiles\objtreelite_runtime.dir\
c_api\c_api_runtime.cc.obj -c D:\treelite\src\c_api\c_api_runtime.cc
[ 91%] Building CXX object src/CMakeFiles/objtreelite_runtime.dir/predictor/predictor.cc.obj
cd /d D:\treelite\build\src && C:\PROGRA~2\mingw64\bin\G__~1.EXE -DTREELITE_OPENMP_SUPPORT @CMakeFiles/objtreelite_runtime.dir/includes_CXX.rsp -O3 -DNDEBUG -funroll-loops -fopenmp -std=g
nu++14 -MD -MT src/CMakeFiles/objtreelite_runtime.dir/predictor/predictor.cc.obj -MF CMakeFiles\objtreelite_runtime.dir\predictor\predictor.cc.obj.d -o CMakeFiles\objtreelite_runtime.dir\
predictor\predictor.cc.obj -c D:\treelite\src\predictor\predictor.cc
D:\treelite\src\predictor\predictor.cc: In member function 'void* treelite::predictor::SharedLibrary::LoadFunction(const char*) const':
D:\treelite\src\predictor\predictor.cc:189:64: error: invalid static_cast from type 'FARPROC' {aka 'long long int (*)()'} to type 'treelite::predictor::SharedLibrary::FunctionHandle' {aka
 'void*'}
   return static_cast<SharedLibrary::FunctionHandle>(func_handle);
                                                                ^
In file included from D:\treelite\src\predictor\predictor.cc:20:
D:\treelite\src\predictor\thread_pool/thread_pool.h: In instantiation of 'void treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::SetAffinity() [with InputToken = {ano
nymous}::InputToken; OutputToken = {anonymous}::OutputToken; TaskContext = treelite::predictor::Predictor]':
D:\treelite\src\predictor\thread_pool/thread_pool.h:55:7:   required from 'treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::ThreadPool(int, const TaskContext*, treel
ite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::TaskFunc) [with InputToken = {anonymous}::InputToken; OutputToken = {anonymous}::OutputToken; TaskContext = treelite::pre
dictor::Predictor; treelite::predictor::ThreadPool<InputToken, OutputToken, TaskContext>::TaskFunc = void (*)(treelite::predictor::SpscQueue<{anonymous}::InputToken>*, treelite::predictor
::SpscQueue<{anonymous}::OutputToken>*, const treelite::predictor::Predictor*)]'
D:\treelite\src\predictor\predictor.cc:374:6:   required from here
D:\treelite\src\predictor\thread_pool/thread_pool.h:88:53: error: invalid conversion from 'std::thread::native_handle_type' {aka 'long long unsigned int'} to 'HANDLE' {aka 'void*'} [-fper
missive]
       SetThreadAffinityMask(thread_[i].native_handle(), (1ULL << core_id));
                             ~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from C:/PROGRA~2/mingw64/x86_64-w64-mingw32/include/windows.h:70,
                 from D:\treelite\src\predictor\thread_pool/thread_pool.h:15,
                 from D:\treelite\src\predictor\predictor.cc:20:
C:/PROGRA~2/mingw64/x86_64-w64-mingw32/include/winbase.h:1162:61: note:   initializing argument 1 of 'DWORD_PTR SetThreadAffinityMask(HANDLE, DWORD_PTR)'
   WINBASEAPI DWORD_PTR WINAPI SetThreadAffinityMask (HANDLE hThread, DWORD_PTR dwThreadAffinityMask);
                                                      ~~~~~~~^~~~~~~
mingw32-make.exe[2]: *** [src\CMakeFiles\objtreelite_runtime.dir\build.make:94: src/CMakeFiles/objtreelite_runtime.dir/predictor/predictor.cc.obj] Error 1
mingw32-make.exe[2]: Leaving directory 'D:/treelite/build'
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:250: src/CMakeFiles/objtreelite_runtime.dir/all] Error 2
mingw32-make.exe[1]: Leaving directory 'D:/treelite/build'
mingw32-make.exe: *** [Makefile:138: all] Error 2
```

--- a/include/treelite/predictor.h
+++ b/include/treelite/predictor.h
@@ -71,8 +71,13 @@ union Entry {
 
 class SharedLibrary {
  public:
+#ifdef _WIN32
+  using LibraryHandle = HMODULE;
+  using FunctionHandle = FARPROC;
+#else  // _WIN32
   using LibraryHandle = void*;
   using FunctionHandle = void*;
+#endif  // _WIN32
   SharedLibrary();
   ~SharedLibrary();
   void Load(const char* libpath);

--- a/include/treelite/predictor.h
+++ b/include/treelite/predictor.h
@@ -16,6 +16,10 @@
 #include <mutex>
 #include <cstdint>
 
+#ifdef _WIN32
+#include <windows.h>
+#endif  // _WIN32
+
 namespace treelite {
 namespace predictor {
 

--- a/include/treelite/predictor.h
+++ b/include/treelite/predictor.h
@@ -17,6 +17,7 @@
 #include <cstdint>
 
 #ifdef _WIN32
+#define NOMINMAX
 #include <windows.h>
 #endif  // _WIN32
 

--- a/src/predictor/thread_pool/thread_pool.h
+++ b/src/predictor/thread_pool/thread_pool.h
@@ -81,12 +81,7 @@ class ThreadPool {
 
   inline void SetAffinity() {
 #ifdef _WIN32
-    /* Windows */
-    SetThreadAffinityMask(GetCurrentThread(), 0x1);
-    for (int i = 0; i < num_worker_; ++i) {
-      const int core_id = i + 1;
-      SetThreadAffinityMask(thread_[i].native_handle(), (1ULL << core_id));
-    }
+    /* Do not set thread affinity on Windows */
 #elif defined(__APPLE__) && defined(__MACH__)
 #include <TargetConditionals.h>
 #if TARGET_OS_MAC == 1

